types/got-resume/index.d.ts

```diff
@@ -1,147 +1,434 @@
-        /** Function called with progress. */
+        /**
+         * Handler for `progress` event.
+         */
         onProgress?(progress: Progress): void;
-        /** Function called with HTTP response. */
-        onResponse?(res: any): void;
-        /** Promise implementation to use (default: Bluebird v2). */
-        Promise?: typeof Promise;
+
+        /**
+         * Handler for `response` event.
+         */
+        onResponse?(res: IncomingMessage): void;
+
+        /**
+         * Promise implementation to use.
+         *
+         * @default Bluebird v2
+         */
+        Promise?: any;
     }
 
-    class CancelError extends globalThis.Error {
-        name: 'GotResumeCancelError';
+    type TransferStream = {
+        /**
+         * Calling `.cancel()` will abort the transfer and cause the stream to emit an `error`
+         * event with a `CancelError`.
+         *
+         * If the transfer is complete before `.cancel()` is called, no `error` event will be
+         * emitted.
+         *
+         * If `options.pre` function is supplied and `.cancel()` is called while `options.pre` is
+         * running, `.cancel()` method on the promise returned by `options.pre` will be called if
+         * it exists. Otherwise the transfer will abort once the promise resolves.
+         */
+        cancel(): void;
+
+        /**
+         * Parent `Transfer` object.
+         */
+        transfer: Transfer;
+
+        /**
+         * Emitted with a `TransferError` on stream when transfer fails and has exhausted retries.
+         */
+        addListener(
+            event: "error",
+            listener: (error: TransferError | CancelError | globalThis.Error) => any,
+        ): TransferStream;
+        /**
+         * Emitted when transfer completes.
+         *
+         * **NB** Is also emitted after error event if transfer fails.
+         */
+        addListener(event: "end", listener: () => any): TransferStream;
+        /**
+         * Emitted when data received.
+         */
+        addListener(event: "progress", listener: (progress: Progress) => any): TransferStream;
+        /**
+         * Emitted when first HTTP request made to server.
+         *
+         * **NB** Not emitted again for each retry HTTP request. You cannot abort the transfer
+         * with `request.abort()` as the request may be finished if a retry has happened.
+         *
+         * Useful for e.g. determining length of transfer.
+         *
+         * @example
+         * import gotResume = require('got-resume');
+         *
+         * const stream = gotResume('http://google.com/');
+         * stream.on('response', res => console.log('Length: ', stream.transfer.length));
+         */
+        addListener(event: "response", listener: (res: IncomingMessage) => any): TransferStream;
+
+        on(event: "error", listener: (error: TransferError | CancelError | globalThis.Error) => any): TransferStream;
```

Comment:
Seems like you could use an event map to simplify redefining all these callbacks for each method.
Similar to how `WorkerEventMap` is used in `Worker`

```ts
interface TransferStreamArgsMap {
  error: [error: TransferError | CancelError | globalThis.Error];
  end: [];
  progress: [progress: Progress];
  response: [res: IncomingMessage];
}

on<E extends keyof TransferStreamArgsMap>(event: E, listener: (...args: TransferStreamArgsMap[K]) => any): TransferStream;
```

Reply:
Tried it out, this has the downside of constraining the TS version to >= 4.0. And somehow it seems to trigger a bug in TS 4.3 (which is rc). Because of this I'd like to leave it as is, even if it's repetitive.